### PR TITLE
feat(fvm): Support v3 config file

### PR DESCRIPTION
**Pull request by provokateurin**
## Changes

fvm v3 has a new config file at `.fvmrc` and uses a slightly different syntax in the config file.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/fvm/extract.spec.ts [hunk 1]
```diff
@@ -36,6 +36,19 @@ describe('modules/manager/fvm/extract', () => {
           packageName: 'flutter/flutter',
         },
       ]);
+
+      res = extractPackageFile(
```
Comment (viceice) [inline lib/modules/manager/fvm/extract.spec.ts:40]: use a new test

File: lib/modules/manager/fvm/index.ts [hunk 2]
```diff
@@ -6,6 +6,6 @@ export { extractPackageFile } from './extract';
 export const supportedDatasources = [FlutterVersionDatasource.id];
 
 export const defaultConfig = {
-  fileMatch: ['(^|/)\\.fvm/fvm_config\\.json$'],
+  fileMatch: ['(^|/)\\.fvm(/fvm_config\\.json|rc)$'],
```
Comment (viceice) [inline lib/modules/manager/fvm/index.ts:9]:
```suggestion
  fileMatch: ['(^|/)\\.fvm(?:/fvm_config\\.json|rc)$'],
```

File: lib/modules/manager/fvm/extract.ts [hunk 3]
```diff
@@ -19,22 +20,26 @@ export function extractPackageFile(
...
-  if (!fvmConfig.flutterSdkVersion) {
-    logger.debug(
-      { contents: fvmConfig },
-      'FVM config does not have flutterSdkVersion specified',
-    );
-    return null;
-  } else if (!is.string(fvmConfig.flutterSdkVersion)) {
-    logger.debug({ contents: fvmConfig }, 'flutterSdkVersion must be a string');
-    return null;
+  for (const version of [fvmConfig.flutterSdkVersion, fvmConfig.flutter]) {
```
Comment (viceice) [inline lib/modules/manager/fvm/extract.ts:23]: extract a helper function instead of using a loop

File: lib/modules/manager/fvm/extract.ts [hunk 4]
```diff
@@ -3,38 +3,43 @@ import { logger } from '../../../logger';
...
-interface FvmConfig {
-  flutterSdkVersion: string;
-}
-
 export function extractPackageFile(
   content: string,
   packageFile: string,
 ): PackageFileContent | null {
-  let fvmConfig: FvmConfig;
+  let fvmConfig: any;
```
Comment (viceice) [inline lib/modules/manager/fvm/extract.ts:10]: no any! use zod Schema

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/fvm/extract.ts (hunk 4 above)
Comment (provokateurin) [reply lib/modules/manager/fvm/extract.ts:10]: Sorry, can you elaborate? I have little experience Typescript and none with zod.

File: lib/modules/manager/fvm/index.ts [hunk 5]
```diff
@@ -6,6 +6,6 @@ export { extractPackageFile } from './extract';
 export const supportedDatasources = [FlutterVersionDatasource.id];
 
 export const defaultConfig = {
-  fileMatch: ['(^|/)\\.fvm/fvm_config\\.json$'],
+  fileMatch: ['(^|/)\\.fvm(?:/fvm_config\\.json|rc)$'],
```
Comment (secustor) [inline lib/modules/manager/fvm/index.ts:9]: Split this in two different regexes for better readability:
```suggestion
  fileMatch: [
    '(^|/)\\.fvm/fvm_config\\.json$',
    '(^|/)\\.fvmrc$'
   ],
```

File: lib/modules/manager/fvm/extract.ts (hunk 4 above)
Comment (secustor) [reply lib/modules/manager/fvm/extract.ts:10]: Take a look at this implementation

https://github.com/renovatebot/renovate/blob/8fb2a6b9d24ba40e6e3482fcde542b4529dbeae5/lib/modules/manager/ocb/extract.ts

https://github.com/renovatebot/renovate/blob/8fb2a6b9d24ba40e6e3482fcde542b4529dbeae5/lib/modules/manager/ocb/schema.ts

**Review by secustor — CHANGES_REQUESTED** [review]

File: lib/modules/manager/fvm/extract.ts (hunk 4 above)
Comment (HonkingGoose) [reply lib/modules/manager/fvm/extract.ts:10]: We also have some developer documentation for Zod here:

https://github.com/renovatebot/renovate/blob/main/docs/development/zod.md

**Review by rarkins — COMMENTED** [review]
@viceice @secustor I think we're stuck on zod here. The current fvm doesn't use it, so we aren't really going backwards by not using it for this PR, are we?

**Comment by viceice** [comment]
> @viceice @secustor I think we're stuck on zod here. The current fvm doesn't use it, so we aren't really going backwards by not using it for this PR, are we?

in current state we're going back because of `any` usage instead of types.

**Comment by provokateurin** [comment]
https://github.com/renovatebot/renovate/pull/28665 Ok so someone did it already. Thanks @Laennart! (I was about to revive this PR)

**Pull request closed without merging**